File: input_reader.c
#include "structs.h"
#include "input_reader.h"

void read_main_data(char* filename, int* videos_nb, int* cache_servers_nb, int* endpoints_nb, int* requests_desc_nb, int* cache_size) {
	FILE *f = fopen(filename, "r");
	
	// read input nbs
	fscanf(f, "%d %d %d %d %d", videos_nb, endpoints_nb, requests_desc_nb, cache_servers_nb, cache_size);
	
	fclose(f);
}

void read_input(char* filename, Video** videos, CacheServer** cache_servers, Endpoint** endpoints, int videos_nb, int cache_servers_nb, int endpoints_nb, int requests_desc_nb, int cache_size) {
	FILE *f = fopen(filename, "r");

	// pass first line
	char ignore[255];
	fgets(ignore, 255, f);
	
	// fill in video data
	int i;
	for (i = 0; i < videos_nb; i++) {
		Video* v = (Video*)malloc(sizeof(Video));
		v->uid = i;
		fscanf(f, "%d", &(v->size));
		videos[i] = v;
	}
	
	// fill in cache servers data
	for (i = 0; i < cache_servers_nb; i++) {
		CacheServer* c = (CacheServer*)malloc(sizeof(CacheServer));
		c->uid = i;
		c->size = cache_size;
		c->connectedEndpoints = (Endpoint**)calloc(endpoints_nb, sizeof(Endpoint));
		cache_servers[i] = c;
	}
	
	// fill in endpoints data
	for (i = 0; i < endpoints_nb; i++) {
		Endpoint* e = (Endpoint*)malloc(sizeof(Endpoint));
		e->uid = i;
		fscanf(f, "%d %d", &(e->datacenterLatency), &(e->nbOfCacheServers));
		e->cacheServersLatencies = (int*)malloc(sizeof(int) * cache_servers_nb);
		int j = 0;
		for (j = 0; j < cache_servers_nb; j++) {
			e->cacheServersLatencies[j] = -1;
		}
		for (j = 0; j < e->nbOfCacheServers; j++) {
			int index, value;
			fscanf(f, "%d %d", &index, &value);
			e->cacheServersLatencies[index] = value;
			cache_servers[index]->connectedEndpoints[e->uid] = e;
		}
		e->video_requests = (int*)calloc(videos_nb, sizeof(int));
		endpoints[i] = e;
	}
	
	// get request descriptions data
	for (i = 0; i < requests_desc_nb; i++) {
		int endpoint, video, requests_nb;
		fscanf(f, "%d %d %d", &video, &endpoint, &requests_nb);
		endpoints[endpoint]->video_requests[video] = requests_nb;
	}
	
	fclose(f);
}

void print_video(Video* v) {
	printf("Video #%d: %dMB\n", v->uid, v->size);
}

void print_endpoint(Endpoint* e, int videos_nb) {
	printf("Endpoint #%d: %dms datacenter latency, connected to %d cache servers:\n", e->uid, e->datacenterLatency, e->nbOfCacheServers);
	int i;
	for (i = 0; i < e->nbOfCacheServers; i++) {
		printf("Cache %d: latency = %dms\n", i, e->cacheServersLatencies[i]);
	}
	for (i = 0; i < videos_nb; i++) {
		if (e->video_requests[i] == 0) continue;
		printf("%d video requests from video %d to endpoint %d\n", e->video_requests[i], i, e->uid);
	}
}

void print_cache_server(CacheServer* c) {
	printf("Cache server #%d: %dMB\n", c->uid, c->size);
}

File: input_reader.py
from structures import *

def read_input(filename):
    # open file in reading mode
    f = open(filename, 'r')

    # get base data
    videos_nb, endpoints_nb, requests_desc_nb, cache_servers_nb, cache_servers_size = [int(x) for x in f.readline().split()]

    # get detailed data
    videos, endpoints, cache_servers = [], [], []
    videos_sizes = [int(x) for x in f.readline().split()]
    # videos
    for i in range(videos_nb):
        videos.append(Video(i, videos_sizes[i]))
    # cache servers
    for i in range(cache_servers_nb):
        cache_servers.append(CacheServer(i, cache_servers_size))
    # endpoints
    for i in range(endpoints_nb):
        datacenter_latency, caches_nb = [int(x) for x in f.readline().split()]
        cache_servers_latencies = []
        for j in range(caches_nb):
            cache_uid, cache_latency = [int(x) for x in f.readline().split()]
            cache_servers_latencies.append((cache_servers[cache_uid], cache_latency))

        e = Endpoint(i, datacenter_latency, cache_servers_latencies)
        endpoints.append(e)
        for cache, _ in cache_servers_latencies:
            cache.connected_endpoints.append(e)
    # video requests
    for i in range(requests_desc_nb):
        video_uid, endpoint_uid, request_nb = [int(x) for x in f.readline().split()]
        endpoints[endpoint_uid].video_requests.append((video_uid, request_nb))

    # close file
    f.close()

    # return data
    return videos, cache_servers
